text/0000-per-project-config.md

```diff
@@ -0,0 +1,87 @@
+
+We currently only support Notion configuration on a per-user basis in the `hooks.toml` file. This file allows users to change the URLs that tools are downloaded from. However, since a user could be working on both an open-source project (that uses the public registry) and an internal project (that uses a private registry), we instead want to allow the users to specify these settings on a per-project basis.
+
+Additionally, we currently already support per-project settings for the platform (node, npm, and yarn versions), specified in the `toolchain` key inside of `package.json`. Multiple users have requested an alternate method in a separate file, for various reasons (see https://github.com/notion-cli/notion/issues/282 for some discussion of one use-case). As we are already looking at adding per-project hooks to a separate file, that file should also support specifying the platform information.
+
+# Pedagogy
+[pedagogy]: #pedagogy
+
+We need to document the format of the `.notionrc.json` file, which will mostly match the features available in the existing `hooks.toml`, however it will also include the additional settings for `toolchain` similar to . Additionally, we need to document and explain the precedence order for when multiple configs are specified (e.g. user-level _and_ project-level).
+
+Additionally, for consistency of configuration, we should update the user-wide config stored in `~/.notion` from `hooks.toml` to also be named `.notionrc.json`, so that it's immediately obvious that it contains the same settings as the project-level configuration. This will be a breaking change that requires users to translate their settings from the old format to the new.
+
+This feature will primarily be used for more advanced use-cases, as the default public repositories and `package.json` settings should cover the majority of open-source projects.
+
+# Details
+[details]: #details
+
+## User-level configuration file
+
+We will need to update the `HookConfig` to parse the user-level `.notionrc.json`, if it exists. We should also allow the user to specify the user platform (`node`, `npm`, and `yarn`) in that file. To that end, we can likely merge the existing `~/.notion/tools/user/platform.json` into `~/.notion/.notionrc.json` as our source of truth for the user-default selected versions. This may also allow us to simplify some of the logic around determining the currently active platform, as we can merge all of the available configs into a single result once, and then use that to execute a tool.
+
+## Configuration Precedence
+
+When configurations are specified in multiple places, we need to deep merge them in a specified order so that we get a single configuration to work from. The order will be defined generally as more specific configurations having precedence over more general configurations. This means:
+
+- Per-project configuration beats per-user configuration
+- `.notionrc.json` platform spec beats `package.json` platform spec
```

Comment:
Do we still need the `"toolchain"` in `package.json` if there can be the same info in the same format in `.notionrc.json`? It seems too complicated to have two possible configurations in the same project, with differing precedence.

I'm thinking that part of the multi-stage implementation could be to deprecate the `package.json` config and move everything into `.notionrc.json`, to have one source of truth in the project.

Reply:
I think there is definitely still value in supporting the `toolchain` key in `package.json`. Specifically in the (likely very common) case where you don't need to specify any custom hooks, then you don't have to add a file to your project.

It does introduce some complexity in technically allowing you to have different configurations in the same project, but given that `.notionrc.json` will always be a more advanced configuration, I think the benefits of easier onboarding in the simple case outweight the negatives of the more complicated advanced case. We could also mitigate that by including some warnings when performing the merge that the user has two different values specified and that they should remove the `package.json` one since it isn't being used anyway.

Reply:
I personally am a fan of consolidating things into `.notionrc.json`. Though if we keep `toolchain`, notion should just error out if the values are different than the ones specified in `.notionrc.json`. That seems like the easiest to implement and doesn't put the users into a weird state.